os/mm/mm_heap/mm_heapinfo_parse_heap.c

```diff
@@ -64,11 +64,109 @@
 #define HEAPINFO_INT INT16_MAX
 #define HEAPINFO_NONSCHED (INT16_MAX - 1)
 
+/****************************************************************************
+ * Public data
+ ****************************************************************************/
+
+typedef enum
+{
+	TYPE_CORRUPTED,
+	TYPE_OVERFLOWED,
+} node_type_t;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static void dump_node(struct mm_allocnode_s *node, node_type_t type)
+{
+	if (type == TYPE_CORRUPTED) {
+		mdbg("CORRUPTED NODE: addr = 0x%08x size = %u preceding size = %u\n", node, node->size, node->preceding & ~MM_ALLOC_BIT);
+	} else if (type == TYPE_OVERFLOWED) {
+		mdbg("OVERFLOWED NODE: addr = 0x%08x size = %u type = %c\n", node, node->size, node->preceding & MM_ALLOC_BIT ? 'A':'F');
+	}
+#ifdef CONFIG_DEBUG_MM_HEAPINFO
+	mdbg("Node owner pid = %u, allocated by code at addr = 0x%08x\n", node->pid, node->alloc_call_addr);
+#endif
+}
+
 /****************************************************************************
  * Public Functions
  ****************************************************************************/
 
 #ifdef CONFIG_DEBUG_MM_HEAPINFO
+
+/****************************************************************************
+ * Name: heapinfo_check_corruption
+ *
+ * Description:
+ *   This function walk through heap and prints information about corrupt node.
+ ****************************************************************************/
+int heapinfo_check_corruption(struct mm_heap_s *heap)
+{
+	struct mm_allocnode_s *node;
+	struct mm_allocnode_s *prev;
+	struct mm_allocnode_s *next;
+#if CONFIG_KMM_REGIONS > 1
+	int region;
+#else
+#define region 0
+#endif
+
+	DEBUGASSERT(heap);
+
+	/* Visit each region */
+
+#if CONFIG_KMM_REGIONS > 1
+	for (region = 0; region < heap->mm_nregions; region++)
+#endif
+	{
+		/* Visit each node in the region
+		 * Retake the semaphore for each region to reduce latencies
+		 */
+
+		mm_takesemaphore(heap);
+
+		prev = NULL;
+		node = heap->mm_heapstart[region];
+		next = (struct mm_allocnode_s *)((char *)node + node->size);
+
+		for (; node < heap->mm_heapend[region]; prev = node, node = next, next = (struct mm_allocnode_s *)((char *)next + next->size)) {
+			if (prev && prev->size != (node->preceding & ~MM_ALLOC_BIT)) {
```

Comment:
To understand easier and faster, how about using some macros?
```
#define MM_PREV_NODE_SIZE(x)  ((x)->preceding & ~MM_ALLOC_BIT)
#define IS_ALLOCATED_NODE(x)   ((x)->preceding & MM_ALLOC_BIT)
#define IS_FREE_NODE(x)               (!IS_ALLOCATED_NODE(x))
```

Reply:
Done